contracts/rollover/CrossCurrencyRollover.sol

```diff
@@ -0,0 +1,437 @@
+        // the swap will revert if swappedAmount < minAmountOut
+        uint256 swappedAmount = _swapExactInputSingle(
+            newLoanTerms.payableCurrency,
+            oldLoanData.terms.payableCurrency,
+            newLoanTerms.principal,
+            swapParams.minAmountOut,
+            swapParams.poolFeeTier,
+            address(this)
+        );
+
+        address oldLender = ILoanCore(loanCore).lenderNote().ownerOf(oldLoanId);
+
+        // get fee snapshot from fee controller
+        (feeSnapshot) = feeController.getFeeSnapshot();
+
+        // calculate settle amounts
+        // get interest amount due
+        uint256 interestAmount = InterestCalculator.getProratedInterestAmount(
+            oldLoanData.balance,
+            oldLoanData.terms.interestRate,
+            oldLoanData.terms.durationSecs,
+            oldLoanData.startDate,
+            oldLoanData.lastAccrualTimestamp,
+            block.timestamp
+        );
+
+        // calculate the repay amount to settle the original loan
+        repayAmount = oldLoanData.terms.principal + interestAmount;
+
+        OriginationLibrary.RolloverAmounts memory amounts = rolloverAmounts(
+            oldLoanData.terms.principal,
+            interestAmount,
+            swappedAmount,
+            lender,
+            oldLender,
+            0,
+            0
+        );
+
+        // borrower owes
+        if (amounts.needFromBorrower > 0) {
+            IERC20(oldLoanData.terms.payableCurrency).safeTransferFrom(borrower, address(this), amounts.needFromBorrower);
+        }
+
+        // amount = original loan repayment amount - leftover principal
+        uint256 amount = amounts.amountFromLender - amounts.leftoverPrincipal;
+
+        if (swappedAmount > amount) {
+            uint256 remainingFunds = swappedAmount - amount;
+
+            // if funds remain, send to the borrower
+            IERC20(oldLoanData.terms.payableCurrency).safeTransfer(borrower, remainingFunds);
+        }
+    }
+
+    /**
+     * @notice Repays the original loan.
+     *
+     * @param borrower                     The address of the borrower.
+     * @param payableCurrency              Payable currency for the loan terms.
+     * @param borrowerNoteId               ID of the borrowerNote for the loan to be repaid.
+     * @param repayAmount                  The amount to be repaid to the old loan.
+     */
+    function _repayLoan(
+        address borrower,
+        IERC20 payableCurrency,
+        uint256 borrowerNoteId,
+        uint256 repayAmount
+    ) internal {
+        // pull BorrowerNote from the caller so that this contract receives collateral upon original loan repayment
+        // borrower must approve this withdrawal
+        IPromissoryNote(borrowerNote).transferFrom(borrower, address(this), borrowerNoteId);
+
+        // approve LoanCore to take the total settled amount
+        ILoanCore loanCoreInterface = ILoanCore(loanCore);
```

Comment:
this line is unnecessary, just do address(loanCore) in the safeApprove line. In the OC it is already an interface